src/Kernel/Magic/DebugMagic.cs

```diff
@@ -0,0 +1,246 @@
+        {
+            public DebugStateDumper (QuantumSimulator qsim) : base(qsim)
+            {
+
+            }
+
+            public override bool Callback(uint idx, double real, double img)
+            {
+                if (_data == null) throw new Exception("Expected data buffer to be initialized before callback, but it was null.");
+                _data[idx] = new Complex(real, img);
+                return true;
+            }
+            private Complex[]? _data = null;
+            public Complex[] getAmplitudes() {
+                var count = this.Simulator.QubitManager.GetAllocatedQubitsCount();
+                _data = new Complex[1 << ((int)count)];
+                var result = base.Dump();
+                    return _data;
+            }
+        }
+
+    public class DebugMagic : AbstractMagic
+    {
+        public DebugMagic(
+                ISymbolResolver resolver, IConfigurationSource configurationSource, IShellRouter router, IShellServer shellServer,
+                ILogger<DebugMagic> logger
+        ) : base(
+            "debug",
+            new Documentation
+            {
+                Summary = "TODO"
+            })
+        {
+            this.SymbolResolver = resolver;
+            this.ConfigurationSource = configurationSource;
+            this.shellServer = shellServer;
+            this.logger = logger;
+            router.RegisterHandler("iqsharp_debug_request", this.HandleStartMessage);
+            router.RegisterHandler("iqsharp_debug_advance", this.HandleAdvanceMessage);
+        }
+
+        private const string ParameterNameOperationName = "__operationName__";
+        private ConcurrentDictionary<Guid, ManualResetEvent> sessionAdvanceEvents
+            = new ConcurrentDictionary<Guid, ManualResetEvent>();
+        private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+        private IShellServer shellServer;
+        private ILogger<DebugMagic> logger;
+
+        /// <summary>
+        ///      The symbol resolver used by this magic command to find
+        ///      operations or functions to be simulated.
+        /// </summary>
+        public ISymbolResolver SymbolResolver { get; }
+
+        /// <summary>
+        ///     The configuration source used by this magic command to control
+        ///     simulation options (e.g.: dump formatting options).
+        /// </summary>
+        public IConfigurationSource ConfigurationSource { get; }
+
+        /// <inheritdoc />
+        public override ExecutionResult Run(string input, IChannel channel) =>
+            RunAsync(input, channel).Result;
+
+        private async Task HandleStartMessage(Message message) => await Task.Run(() =>
+        {
+            var content = (message.Content as UnknownContent);
+            var session = content?.Data?["debug_session"];
+            if (session == null)
+            {
+                // TODO: Communicate error back here!
```

Comment:
It would be good to at least log an error here, for example:
```
this.logger.LogError("Missing debug_session field in iqsharp_debug_request message.");
```
Then you could probably remove the TODO. This is a pretty unexpected case, so I don't think any further error handling would be needed.

Reply:
added that line to log the error